---
title: "Filter of PED and MAP files"
author: "FiltPeds Tool"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
params:
  type: NA
  probes: NA
  list: NA
  m1: NA
  m2: NA
  p: NA
---

```{r setup, include=FALSE, warning=FALSE, message=FALSE}
library(knitr)
library(dplyr)
opts_chunk$set(echo = FALSE)
```

## Usage of FiltPeds Tool

MAP and PED files are very widely used for analyzing genotypic data. Usually, from the hundreds of thousands of variants, users are only interested in analyzing specific genes that represent a small percentage of the whole data. The **FiltPEDs Tool** is designed to generate new MAP and PED files with selected variants, based on the gene where it occurs. It is designed for data obtained from the **Infinium Global Screening Array-24 v1.0 BeadChip**, of Illumina, customized for the Instituto Nacional de Medicina Gen√≥mica, with 669672 variants in total.

The **Annotation** module contains the full annotated probes of the microarray, with Ensembl ID, gene symbol from the HUGO Gene Nomenclature Committee database and variants ID from the dbSNP database. Some variants may be associated more than one gene to them if the variant has implications on genes downstream or upstream. Some variants do not have genes associated to them if the variant occurs on introns or intergenic sites. Some variants are not found in the dbSNP database, thus they do not have an *rs* identifier; in those cases, an artificial variant id was created with the format `rs0067 + chromosome + position` for further analysis in PLINK or similar software; variants in chromosomes *X*, *Y*, *XY* and *MT*, take the value of 0 in `chromosome`. Users can filter the original annotation dataset by probe name, probe-variant name, chromosome, position, variant ID, Ensembl ID or gene symbol; filtered annotation dataset can be download as a CSV file.

The **Select filters** module allows filtering the annotation database with multiple values of variant IDs, Ensembl IDs or gene symbols. First, select the type of filter you want to apply; then upload a TXT/CSV file containing the values to filter, each value in a single line, such as the following example:
```
PRDM2
ADH6
STK19
BANK1
SORCS2
BRCA2
SLC2A9
ANAPC1
```
You will see a summary table with the genes and variants that your new annotation dataset contains, and the values that do not have any match.

The **Filter PED** module allows getting a MAP and a PED file, with the filtered dataset obtained in the previous module. Upload a MAP with the following format:
```
13	rs276174824	0	32911095
16	rs1042014	0	28618318
4	IlmnSeq_rs10032900	0	7491757
4	GSA-rs112438204	0	146015170
1	1KG_1_14106394	0	14106394
11	rs113631738	0	119024863
4	rs9997426	0	7399225
8	rs12200314	0	52481
```
where the first column is the chromosome, the second the probe-variant name (as is in the annotation dataset), the third the genetic distance and the fourth the physical position. Then, upload a PED file with the following format:
```
FAM1	NA06985	0	0	1	1	A	T	T	T	G	G	C	C	A	T	T	T	G	G	C	C
FAM1	NA06991	0	0	1	1	C	T	T	T	G	G	C	C	C	T	T	T	G	G	C	C
FAM1	NA06993	0	0	1	1	C	T	T	T	G	G	C	T	C	T	T	T	G	G	C	T
FAM2	NA06994	0	0	1	1	C	T	T	T	G	G	C	C	C	T	T	T	G	G	C	C
FAM2	NA07000	0	0	2	1	C	T	T	T	G	G	C	T	C	T	T	T	G	G	C	T
FAM2	NA07019	0	0	1	1	C	T	T	T	G	G	C	C	C	T	T	T	G	G	C	C
FAM3	NA07022	0	0	2	1	C	T	T	T	G	G	0	0	C	T	T	T	G	G	0	0
FAM3	NA07029	0	0	1	1	C	T	T	T	G	G	C	C	C	T	T	T	G	G	C	C
FAM3	NA07056	0	0	0	2	C	T	T	T	A	G	C	T	C	T	T	T	A	G	C	T
FAM3	NA07345	0	0	1	1	C	T	T	T	G	G	C	C	C	T	T	T	G	G	C	C
```
where the first column is the family ID, the second the sample ID,  the third the paternal ID, the fourth the maternal ID, the fifth the sex (1=male; 2=female; other=unknown), the sixth the affection (0=unknown; 1=unaffected; 2=affected), and from the seventh the genotypes. *MAP and PED files must be tab-separated and without headings*. PED file must coincide with the MAP file: the number of columns on PED file must be *6 + 2n*, where *n* is the number of rows in the MAP file. Click on *Ready to filter*. You will see a message indicating the number of variants in the original MAP file and the number in the new one; additionally, you will see a summary of the samples in the PED file.

The **Download files** module allows downloading the generated files: the new MAP file, the new annotation dataset file, and the new PED file.

After you download the generated files, you can upload a new PED, MAP or filtered file.

## Summary table of features

In order to filter MAP and PED files, the filter type used was `r params$type`. Following genes are included in the new files:

```{r filter}
summary_table <- params$probes %>% select(Variant, Ensembl, Gene) %>%
  group_by(Gene) %>%
  summarize(Ensembl = paste(unique(Ensembl), collapse = ", "),
            Variants = n_distinct(Variant)) %>%
  ungroup()
not_found <- setdiff(params$list, 
                     params$probes[[params$type]])
kable(summary_table)
```

Following `r params$type` IDs were not found in annotation dataset:

`r paste(not_found, collapse = ', ')`.

## Summary of MAP file

Original MAP file had `r nrow(params$m1)` variants. New MAP file has `r nrow(params$m2)` variants.

## Summary of PED file

PED file has `r nrow(params$p)` samples:

```{r ped}
sum_ped <- params$p %>%
  transmute("Family ID" = V1, "Sample ID" = V2, "Paternal ID" = V3,
            "Maternal ID" = V4, "Sex" = V5, "Affection" = V6)
kable(sum_ped)
```
